intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,519 @@
+# As you can see, matrix multiply does not check if the contracted dimensions
+# have the same name.
+#
+# Accessors and Reduction
+# -----------------------
+#
+# One can use dimension names to refer to dimensions instead of the positional
+# dimension. These operations also propagate names. Indexing (basic and
+# advanced) has not been implemented yet but is on the roadmap. Using the ``named_imgs``
+# tensor from above, we can do:
+
+output = named_imgs.sum(['C'])  # Perform a sum over the channel dimension
+print(output.names)
+
+img0 = named_imgs.select('N', 0)  # get one image
+print(img0.names)
+
+######################################################################
+# Name inference
+# --------------
+#
+# Names are propagated on operations in a two step process called **name inference**.
+# The two steps are:
+#
+# 1. **Check names**: an operator may perform automatic checks at runtime that
+#    check that certain dimension names must match.
+# 2. **Propagate names**: name inference propagates output names to output tensors.
+#
+# Let's go through the very small example of adding 2 one-dim tensors with no
+# broadcasting.
+
+x = torch.randn(3, names=('X',))
+y = torch.randn(3)
+z = torch.randn(3, names=('Z',))
+
+######################################################################
+# **Check names**: first, we will check whether the names of these two tensors
+# match. Two names match if and only if they are equal (string equality) or at
+# least one is ``None`` (``None``s are essentially a special wildcard name).
+# The only one of these three that will error, therefore, is ``x+z``:
+
+catch_error(lambda: x + z)
+
+######################################################################
+# **Propagate names**: _unify_ the two names by returning the most refined name of
+# the two. With `x + y,  `X` is more refined than `None`.
+
+print((x + y).names)
+
+######################################################################
+# Most name inference rules are straightforward but some of them can have
+# unexpected semantics. Let's go through a few more of them.
+#
+# Broadcasting
+# ------------
+#
+# Named tensors do not change broadcasting behavior; they still broadcast by
+# position. However, when checking two dimensions for if they can be
+# broadcasted, the names of those dimensions must match.
+#
+# Furthermore, broadcasting with named tensors can prevent incorrect behavior.
+# The following code will error, whereas without `names` it would add
+# `per_batch_scale` to the last dimension of `imgs`.
+
+# Automatic broadcasting: expected to fail
+imgs = torch.randn(6, 6, 6, 6, names=('N', 'C', 'H', 'W'))
+per_batch_scale = torch.rand(6, names=('N',))
+catch_error(lambda: imgs * per_batch_scale)
+
+######################################################################
+# How `should` we perform this broadcasting operation along the first
+# dimension? One way, involving names, would be to explicitly initialize

```

Comment:
Let's not call it "explicit" so that we don't confuse it with the "explicit broadcasting by names" section below. Can we change this line to:

"One way, involving names, would be to name the ``per_batch_scale`` tensor such that it matches ``imgs.names``, as shown below."

And then let's change `per_batch_scale_4d`'s names to `('N', 'C', 'H', 'W')` instead of being partially named.

Reply:
Much better 😀